github.com/mattn/go-sqlite3 can't be built as an external dependency

go-sqlite3 contains a header file, `sqlite3-binding.h`, which can be included when go-sqlite3 is built as the local repository, but not when it's built as an external dependency.

```
ERROR: /usr/local/google/home/jayconrod/.cache/bazel/_bazel_jayconrod/558e6c6fca1bec20ec6f31ee3f9f34ec/external/com_github_mattn_go_sqlite3/BUILD.bazel:5:1: error executing shell command: 'bazel-out/local-fastbuild/bin/external/com_github_mattn_go_sqlite3/bazel-out/local-fastbuild/genfiles/external/com_github_mattn_go_sqlite3/cgo_default_library.cgo_codegen.dir/CGoCodeGenFile.params' failed: bash failed: error executing command /bin/bash -c ... (remaining 1 argument(s) skipped): com.google.devtools.build.lib.shell.BadExitStatusException: Process exited with status 1.
/usr/local/google/home/jayconrod/.cache/bazel/_bazel_jayconrod/558e6c6fca1bec20ec6f31ee3f9f34ec/execroot/popular_repos/bazel-out/local-fastbuild/genfiles/external/com_github_mattn_go_sqlite3/cgo_default_library.cgo_codegen.dir/external/com_github_mattn_go_sqlite3/backup.go:10:29: fatal error: sqlite3-binding.h: No such file or directory
 #include <sqlite3-binding.h>
                             ^
compilation terminated.
blaze: Leaving directory `/usr/local/google/home/jayconrod/.cache/bazel/_bazel_jayconrod/558e6c6fca1bec20ec6f31ee3f9f34ec/execroot/popular_repos/'
Target @com_github_mattn_go_sqlite3//:go_default_library failed to build
```

Note that cgo itself is failing, not the compiler used by `cc_library`. We may need to change the directory where cgo runs or adjust the include paths. This adjustment may also be needed for `cc_library`.

--- a/go/private/cgo.bzl
+++ b/go/private/cgo.bzl
@@ -408,6 +408,11 @@ def _setup_cgo_library(name, srcs, cdeps, copts, clinkopts):
   
   # Run cgo on the filtered Go files. This will split them into pure Go files
   # and pure C files, plus a few other glue files.
+  pkg_dir = _pkg_dir(
+      "external/" + REPOSITORY_NAME[1:] if len(REPOSITORY_NAME) > 1 else "",
+      PACKAGE_NAME)
+  copts += ["-I", pkg_dir]
+
   cgo_codegen_name = name + ".cgo_codegen"
   _cgo_codegen_rule(
       name = cgo_codegen_name,
@@ -436,10 +441,6 @@ def _setup_cgo_library(name, srcs, cdeps, copts, clinkopts):
   # Compile C sources and generated files into a library. This will be linked
   # into binaries that depend on this cgo_library. It will also be used
   # in _cgo_.o.
-  pkg_dir = _pkg_dir(
-      "external/" + REPOSITORY_NAME[1:] if len(REPOSITORY_NAME) > 1 else "",
-      PACKAGE_NAME)
-
   platform_copts = select({
       "@io_bazel_rules_go//go/platform:windows_amd64": ["-mthreads"],
       "//conditions:default": ["-pthread"],
@@ -452,7 +453,6 @@ def _setup_cgo_library(name, srcs, cdeps, copts, clinkopts):
       srcs = [cgo_codegen_name],
       deps = cdeps,
       copts = copts + platform_copts + [
-          "-I", pkg_dir,
           "-I", "$(BINDIR)/" + pkg_dir + "/" + cgo_codegen_dir,
           # The generated thunks often contain unused variables.
           "-Wno-unused-variable",

--- a/tests/cgo_sys_hdr/BUILD
+++ b/tests/cgo_sys_hdr/BUILD
@@ -1,0 +1,15 @@
+load("@io_bazel_rules_go//go:def.bzl", "cgo_library", "go_test")
+
+cgo_library(
+    name = "cgo_default_library",
+    srcs = [
+        "foo.go",
+        "sub/foo.h",
+    ],
+)
+
+go_test(
+    name = "go_default_test",
+    srcs = ["foo_test.go"],
+    library = ":cgo_default_library",
+)

--- a/tests/cgo_sys_hdr/foo.go
+++ b/tests/cgo_sys_hdr/foo.go
@@ -1,0 +1,8 @@
+package cgo_sys_hdr
+
+/*
+#include <sub/foo.h>
+*/
+import "C"
+
+var x = int(C.x)

--- a/tests/cgo_sys_hdr/foo_test.go
+++ b/tests/cgo_sys_hdr/foo_test.go
@@ -1,0 +1,9 @@
+package cgo_sys_hdr
+
+import "testing"
+
+func TestFoo(t *testing.T) {
+	if x != 42 {
+		t.Errorf("got %d; want %d", x, 42)
+	}
+}

--- a/tests/cgo_sys_hdr/sub/foo.h
+++ b/tests/cgo_sys_hdr/sub/foo.h
@@ -1,0 +1,1 @@
+const int x = 42;

--- a/tests/popular_repos/popular_repos.bash
+++ b/tests/popular_repos/popular_repos.bash
@@ -39,6 +39,7 @@ targets=(
   @org_golang_x_sys//...
   @org_golang_x_text//...
   @org_golang_x_tools//...
+  @com_github_mattn_go_sqlite3//...
 )
 
 excludes=(
@@ -83,6 +84,10 @@ excludes=(
   -@org_golang_x_tools//go/ssa:go_default_xtest
   -@org_golang_x_tools//refactor/eg:go_default_xtest
 
+  # TODO(#546): deps don't get propagated from cgo_library through go_library
+  # to go_test.
+  -@com_github_mattn_go_sqlite3//:go_default_test
+
   # TODO(#417): several tests fail. Need to investigate and fix.
   -@org_golang_x_tools//cmd/godoc:go_default_xtest
   -@org_golang_x_tools//go/gcimporter15:go_default_xtest
@@ -94,6 +99,11 @@ excludes=(
 
   # fiximports requires working GOROOT, not present in CI.
   -@org_golang_x_tools//cmd/fiximports:go_default_test
+
+  # some sqlite3 tests rely on sources in the library which are filtered out.
+  -@com_github_mattn_go_sqlite3//_example/custom_func:all
+  -@com_github_mattn_go_sqlite3//_example/trace:all
+  -@com_github_mattn_go_sqlite3//_example/vtable:all
 )
 
 case $(uname) in

